sdks/go/pkg/beam/core/sdf/sdf.go

```diff
@@ -13,63 +13,64 @@
-	// restriction's start and end points, and every block of work in a restriction must be claimed.
+	// If the claim fails due to an error, that error is stored and will be automatically emitted
+	// when the RTracker is validated, or can be manually retrieved with GetError.
 	//
 	// This pseudocode example illustrates the typical usage of TryClaim:
 	//
-	// 	pos = position of first block after restriction.start
+	// 	pos = position of first block within the restriction
 	// 	for TryClaim(pos) == true {
 	// 		// Do all work in the claimed block and emit outputs.
-	// 		pos = position of next block
+	// 		pos = position of next block within the restriction
 	// 	}
 	// 	return
 	TryClaim(pos interface{}) (ok bool)
 
-	// GetError returns the error that made this RTracker stop executing, and it returns nil if no
-	// error occurred. If IsDone fails while validating this RTracker, this method will be
-	// called to log the error.
+	// GetError returns the error that made this RTracker stop executing, and returns nil if no
+	// error occurred. This is the error that is emitted if automated validation fails.
 	GetError() error
 
-	// TrySplit splits the current restriction into a primary and residual based on a fraction of the
-	// work remaining. The split is performed along the first valid split point located after the
-	// given fraction of the remainder. This method is called by the SDK harness when receiving a
-	// split request by the runner.
+	// TrySplit splits the current restriction into a primary (currently executing work) and
+	// residual (work to be split off) based on a fraction of work remaining. The split is performed
+	// at the first valid split point located after the given fraction of remaining work.
+	//
+	// For example, a fraction of 0.5 means to split at the halfway point of remaining work only. If
+	// 50% of work is done and 50% remaining, then a fraction of 0.5 would split after 75% of work.
+	//
+	// This method modifies the underlying restriction in the RTracker to reflect the primary. It
+	// then returns a copy of the newly modified restriction as a primary, and returns a new
+	// restriction for the residual. If the split would produce an empty residual (i.e. the only
+	// split point is the end of the restriction), then the returned residual is nil.
 	//
-	// The current restriction is split into two by modifying the current restriction's endpoint to
-	// turn it into the primary, and returning a new restriction tracker representing the residual.
-	// If no valid split point exists, this method returns nil instead of a residual, but does not
-	// return an error. If this method is unable to split due to some error then it returns nil and
-	// an error.
-	TrySplit(fraction float64) (residual interface{}, err error)
+	// If an error is returned, some catastrophic failure occurred and the entire bundle will fail.
+	TrySplit(fraction float64) (primary, residual interface{}, err error)
```

Comment:
IIRC for dynamic splitting, this is the one that requires the Tracker to be concurrency safe? Do we want to declared ahead of time that implementations of RTracker must be threadsafe?

Reply:
Hmm, good point. It's definitely something a use should know, especially for writing future-proof SDFs. It would need to be in all the methods I think, except *maybe* GetError. I'm not completely sure off the top of my head though, I feel like it'll depend on the details of how we do dynamic splitting.

For now I'll add a comment saying to make all methods thread-safe. Later we should add the thread-safe wrapper and recommend using that, and if we can dodge mutexes in some methods then we can implement it there.

Reply:
Technically since GetError is going to be reading from a location that might be concurrently modified, it's required to be thread safe as well.